### PR TITLE
8306021: Add event handler management to EventTarget

### PR DESCRIPTION
**Pull request by mstr2**
This PR adds the following methods to the `EventTarget` interface:
1. `addEventHandler`
2. `removeEventHandler`
3. `addEventFilter`
4. `removeEventFilter`

---------
### Progress
- [x] Change requires CSR request [JDK-8306035](https://bugs.openjdk.org/browse/JDK-8306035) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (3 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 2 [Authors](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8306021](https://bugs.openjdk.org/browse/JDK-8306021): Add event handler management to EventTarget (**Enhancement** - P4)
 * [JDK-8306035](https://bugs.openjdk.org/browse/JDK-8306035): Add event handler management to EventTarget (**CSR**)

### Reviewers
 * [John Hendrikx](https://openjdk.org/census#jhendrikx) (@hjohn - Committer)
 * [Nir Lisker](https://openjdk.org/census#nlisker) (@nlisker - **Reviewer**)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Marius Hanl](https://openjdk.org/census#mhanl) (@Maran23 - Committer)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1090/head:pull/1090` \
`$ git checkout pull/1090`

Update a local copy of the PR: \
`$ git checkout pull/1090` \
`$ git pull https://git.openjdk.org/jfx.git pull/1090/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1090`

View PR using the GUI difftool: \
`$ git pr show -t 1090`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1090.diff">https://git.openjdk.org/jfx/pull/1090.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1090#issuecomment-1509509312)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
If we proceed with this, we will need a CSR that documents the incompatibilities that are mentioned in the thread on openjfx-dev.

/reviewers 3
/csr

**Comment by mstr2** [comment]
This PR assumes that changing the first method signature to the second is a binary-compatible change:
```
1. <E extends Event> void addEventHandler(EventType<E>, EventHandler<E>)
2. <E extends Event> void addEventHandler(EventType<E>, EventHandler<? super E>)
```

To verify this, I've created the following test:
```java
import java.util.function.Consumer;

// Test.java
public class Test {
    public static <T> void test(Consumer<T> c) {
        System.out.println("it works");
    }
}

// Main.java
public class Main {
    public static void main(String[] args) {
        Consumer<String> c = param -> {};
        Test.test(c);
    }
}
```

Note that `Consumer`, like `EventHandler`, only uses the generic type parameter as an _input_ argument.
After compiling both files, I changed the signature of the `test` method to `void test(Consumer<? super T>)`.
Then I made sure to _only_ recompile the `Test.java` file and ran the `Main` program, which succeeded.

**Review by hjohn — APPROVED** [review]
I checked everything, and didn't see any problems; looks good!

**Comment by mstr2** [comment]
I've reverted a change that caused a unit test failure.

**Review by hjohn — APPROVED** [review]

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableColumnBase.java [hunk 1]
```diff
@@ -726,42 +726,42 @@ public final T getCellData(final S item) {
...
-     * Note however that TableColumnBase is <b>not</b> a Node, and therefore no visual
-     * events will be fired on it.
-     *
-     * @param <E> The type of event
-     * @param eventType the type of the events to receive by the handler
-     * @param eventHandler the handler to register
-     * @throws NullPointerException if the event type or handler is null
+     * {@inheritDoc}
+     * <p>
+     * The {@code TableColumnBase} class allows registration of listeners which will be notified when editing occurs.
```
Comment (nlisker) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TableColumnBase.java:737]: The edit events are defined only in the subclasses. Either, if possible, move the events to this class, or clarify that these events are defined in subclasses. `TreeItem` talks more about this - using the cell.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableColumnBase.java [hunk 2]
```diff
@@ -726,42 +726,42 @@ public final T getCellData(final S item) {
...
+     * Note that {@code TableColumnBase} is <b>not</b> a {@link Node}, and therefore no visual events will be fired on it.
```
Comment (nlisker) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TableColumnBase.java:738]: Isn't this true for the other classes here as well, like `Tab`? Might it be worth saying this in the other classes too then, or even on `EventTarget` itself (something like "Note that implementing classes that are not a `Node` don't receive visual events")?
I would also give an example or 2 of what "visual events" are since this term is not defined anywhere that I know of.

**Review by nlisker — COMMENTED** [review]
Looks good overall.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableColumnBase.java (hunk 2 above)
Comment (mstr2) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableColumnBase.java:738]: I don't know what "visual events" are, maybe input events?
Anyway, the added value of this entire javadoc seems to be rather low, I'm inclined to just remove it entirely.

By the way, there's a great documentation of the JavaFX event system at https://docs.oracle.com/javase/8/javafx/events-tutorial/events.htm. It would be really helpful if we could add this to the JavaFX codebase in the future.

Comment (nlisker) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableColumnBase.java:738]: > I don't know what "visual events" are, maybe input events?
I thought something like that too, including scroll and touch events. Maybe @kevinrushforth knows.

> Anyway, the added value of this entire javadoc seems to be rather low, I'm inclined to just remove it entirely.
I don't think I mind removing it. My more general concern is that it's not obvious what events can be registered on different `EventTarget`s.

> By the way, there's a great documentation of the JavaFX event system at https://docs.oracle.com/javase/8/javafx/events-tutorial/events.htm. It would be really helpful if we could add this to the JavaFX codebase in the future.

Yes, I had an idea to link to the tutorials when working on controls documentation (https://bugs.openjdk.org/browse/JDK-8252550). The problem is that we don't have access to them and can't update them if they are out of date. Embedding them is probably too verbose. Linking looks like the better option.

Comment (mstr2) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TableColumnBase.java]: I've removed the documentation from `TableColumnBase` and instead added a paragraph to the class documentation of both `TableColumn` and `TreeTableColumn`.

**Review by nlisker — APPROVED** [review]

**Review by hjohn — APPROVED** [review]

**Review by andy-goryachev-oracle — APPROVED** [review]
I think this is a logical change.
Limited testing on Mac shows no ill effects.

**Review by Maran23 — APPROVED** [review]
LGTM

**Review by andy-goryachev-oracle — APPROVED** [review]

**Comment by andy-goryachev-oracle** [comment]
I think CSR needs to be out of Draft status to move this PR along.

**Comment by mstr2** [comment]
> I think CSR needs to be out of Draft status to move this PR along.

I changed the status to Proposed.

File: modules/javafx.graphics/src/main/java/javafx/concurrent/Service.java [hunk 3]
```diff
@@ -741,74 +741,31 @@ private EventHelper getEventHelper() {
...
-     * in the scene graph, if an event is consumed, it will not continue
-     * dispatching.
-     *
-     * @param <T> the specific event class of the handler
-     * @param eventType the type of the events to receive by the handler
-     * @param eventHandler the handler to register
-     * @throws NullPointerException if the event type or handler is null
-     * @since JavaFX 2.1
-     */
+    @Override
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/concurrent/Service.java:744]: By removing the docs, you lost the `@since` for this method, so you will need to restore it. You might consider also adding `{@inheritDoc}` to avoid copying all of the description

File: modules/javafx.graphics/src/main/java/javafx/concurrent/Task.java [hunk 4]
```diff
@@ -1248,74 +1248,31 @@ private EventHelper getEventHelper() {
...
-     * dispatching.
-     *
-     * @param <T> the specific event class of the handler
-     * @param eventType the type of the events to receive by the handler
-     * @param eventHandler the handler to register
-     * @throws NullPointerException if the event type or handler is null
-     * @since JavaFX 2.1
-     */
+    @Override
     public final <T extends Event> void addEventHandler(
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/concurrent/Task.java:1252]: Same comment in this file about the need to restore the `@since tag`.

File: modules/javafx.graphics/src/main/java/javafx/scene/transform/Transform.java [hunk 5]
```diff
@@ -1909,22 +1909,12 @@ public EventDispatchChain buildEventDispatchChain(EventDispatchChain tail) {
...
-     * {@code TRANSFORM_CHANGED}.
-     * </p>
-     *
-     * @param <T> the specific event class of the handler
-     * @param eventType the type of the events to receive by the handler
-     * @param eventHandler the handler to register
-     * @throws NullPointerException if the event type or handler is null
-     * @since JavaFX 8.0
+     * Currently the only event delivered to a {@code Transform} is the {@code TransformChangedEvent}
+     * with its single type {@code TRANSFORM_CHANGED}.
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/scene/transform/Transform.java:1915]: You need to restore the `@since`

File: modules/javafx.controls/src/main/java/javafx/scene/control/TableColumn.java [hunk 6]
```diff
@@ -74,6 +74,15 @@
...
  *
+ * {@code TableColumn} declares the following {@link CellEditEvent} types, which are received
+ * when a cell is being edited:
+ * <ul>
+ *     <li>{@link #editAnyEvent()}
+ *     <li>{@link #editStartEvent()}
+ *     <li>{@link #editCommitEvent()}
+ *     <li>{@link #editCancelEvent()}
+ * </ul>
+ *
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TableColumn.java:85]: These doc additions are unrelated to the this PR.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableColumn.java [hunk 7]
```diff
@@ -69,6 +69,15 @@
...
  *
+ * {@code TreeTableColumn} declares the following {@link CellEditEvent} types, which are received
+ * when a cell is being edited:
+ * <ul>
+ *     <li>{@link #editAnyEvent()}
+ *     <li>{@link #editStartEvent()}
+ *     <li>{@link #editCommitEvent()}
+ *     <li>{@link #editCancelEvent()}
+ * </ul>
+ *
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableColumn.java:80]: These doc additions are unrelated to the this PR.

**Review by kevinrushforth — CHANGES_REQUESTED** [review]
The spec looks fine except for a few missing `@since` tags that need to be restored, a suggestion on adding `@implSpec` (from Joe), and a couple unrelated doc changes that should be reverted. See inline.

As for the implementation, the event handling methods in the `Dialog` and `Tab` classes are new (although they just delegate to an existing instance of `EventHandlerManager`). Would it be possible to provide a test?

**Review by nlisker — APPROVED** [review]

**Review by hjohn — APPROVED** [review]

**Review by kevinrushforth — APPROVED** [review]
Looks good. I'll Review the CSR and then you can Finalize it.
